### PR TITLE
Add a list of validators that meet the official GTFS requirements

### PR DESCRIPTION
**Pull request by aababilov**
* `BlockTripsWithConsistentTypeValidator`
  * Validates that trips that belong to the same block have consistent route types (e.g., bus does not transfer to rail).
  * Generated notice: `BlockTripsWithInconsistentRouteTypesNotice`.
* `FareAttributeAgencyIdValidator`
  * Checks that `fare_attributes.agency_id` field is defined for every row if there is more than 1 agency in the feed.
  * Generated notice: `MissingRequiredFieldNotice`.
* `FeedHasLanguageValidator`
  * Checks that a feed has language defined in either `feed_info.feed_lang` or `agency.agency_lang`. Note that `feed_lang` is a required field: if there is a `feed_info` entity, it must have `feed_lang` set. `agency_lang`, instead, is optional. The recommended way is to provide `feed_lang` and omit `agency_lang`.
  * Generated notice: `FeedHasNoLanguageNotice`.
* `StationUsageValidator`
  * Checks that every station (`location_type=1`) in `stops.txt` has child platforms (` location_type=0`). Note that a station that has child entrances (`location_type=2`) or generic nodes (`location_type=3`) is also reported.
  * Generated notice: `StationWithoutPlatformsNotice`.
* `StopLatLngRequiredValidator`
  * Checks that conditionally required fields `stop_lat` and `stop_lon` are set depending on `location_type`.
    *  `stop_lat` and `stop_lon` are required for stations (`location_type=0`), stops (`location_type=1`) and entrances (`location_type=2`) and optional for other types.
  * Generated notice: `MissingRequiredFieldError`.
* `StopTimeInSequenceOrderValidator`
  * Validates that fields `arrival_time`, `departure_time` and `shape_dist_traveled` are non-decreasing along each trip.
  * Generated notices:
    * `TripWithOutOfOrderArrivalTimeNotice`
    * `TripWithOutOfOrderDepartureTimeNotice`
    * `TripWithOutOfOrderShapeDistTraveledNotice`
* `TransferFrequencyBasedTripValidator`
  * Validates that trip-to-trip transfers do not involve frequency-based trips. 
  * Generated notice: `TransferForFrequencyBasedTripNotice`.
* `TripShapeDistTraveledValidator`
  * Validates that trips specify `stop_times.shape_dist_traveled` values properly. Please see generated notices for details.
  * Generated notices:
    * `TripWithShapeDistTraveledButNoShapeNotice`
    * `TripWithShapeDistTraveledButNoShapeDistancesNotice`
    * `TripWithPartialShapeDistTraveledNotice`

### CODE REVIEW

File: main/src/main/java/org/mobilitydata/gtfsvalidator/validator/TransferFrequencyBasedTripValidator.java [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+import org.mobilitydata.gtfsvalidator.annotation.GtfsValidator;
+import org.mobilitydata.gtfsvalidator.notice.NoticeContainer;
+import org.mobilitydata.gtfsvalidator.notice.SeverityLevel;
+import org.mobilitydata.gtfsvalidator.notice.ValidationNotice;
+import org.mobilitydata.gtfsvalidator.table.GtfsFrequency;
+import org.mobilitydata.gtfsvalidator.table.GtfsFrequencyTableContainer;
+import org.mobilitydata.gtfsvalidator.table.GtfsTransfer;
+import org.mobilitydata.gtfsvalidator.table.GtfsTransferTableContainer;
+
+/** Validates that trip-to-trip transfers do not involve frequency-based trips. */
```
Comment (asvechnikov2) [inline main/src/main/java/org/mobilitydata/gtfsvalidator/validator/TransferFrequencyBasedTripValidator.java:33]:
```
/**
 * Validates that trip-to-trip transfers do not involve frequency-based trips. 
 * 
 * <p> Generated notice: {@link TransferForFrequencyBasedTripNotice}.
 */
```

File: main/src/main/java/org/mobilitydata/gtfsvalidator/validator/TripShapeDistTraveledValidator.java [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+import org.mobilitydata.gtfsvalidator.notice.SeverityLevel;
+import org.mobilitydata.gtfsvalidator.notice.ValidationNotice;
+import org.mobilitydata.gtfsvalidator.table.GtfsShape;
+import org.mobilitydata.gtfsvalidator.table.GtfsShapeTableContainer;
+import org.mobilitydata.gtfsvalidator.table.GtfsStopTime;
+import org.mobilitydata.gtfsvalidator.table.GtfsStopTimeTableContainer;
+import org.mobilitydata.gtfsvalidator.table.GtfsTrip;
+import org.mobilitydata.gtfsvalidator.table.GtfsTripTableContainer;
+
+/** Validates that trips specify {@code stop_times.shape_dist_traveled} values properly. */
```
Comment (asvechnikov2) [inline main/src/main/java/org/mobilitydata/gtfsvalidator/validator/TripShapeDistTraveledValidator.java:35]:
```
/**
 * Validates that trips specify {@code stop_times.shape_dist_traveled} values properly. 
 * Please see generated notices for details.
 * 
 * <p> Generated notices:
 * <ul>
 *   <li> {@link TripWithShapeDistTraveledButNoShapeNotice}
 *   <li> {@link TripWithShapeDistTraveledButNoShapeDistancesNotice}
 *   <li> {@link TripWithPartialShapeDistTraveledNotice}
 * </ul>
 */
```

**Review by asvechnikov2 — APPROVED** [review]
Thanks! LGTM. 

Hi @isabelle-dr, could you please take a look at the proposed validations if they look good? Do you know if there are any open issues that could be closed by this change?

**Comment by isabelle-dr** [comment]
Thanks for the ping @asvechnikov2! (and.. wow @aababilov 🎉).
I will have a look this week, looking forward to it 😊

**Comment by bdferris-v2** [comment]
The LGTM.  However, you do need to update `RULES.md` with the new validation notices you've added.

**Comment by bdferris-v2** [comment]
I'd also call out that issue #1304 is probably still blocking us from getting a clean acceptance test run.

File: main/src/main/java/org/mobilitydata/gtfsvalidator/validator/FareAttributeAgencyIdValidator.java [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+      GtfsFareAttributeTableContainer fareAttributeTable, GtfsAgencyTableContainer agencyTable) {
+    this.fareAttributeTable = fareAttributeTable;
+    this.agencyTable = agencyTable;
+  }
+
+  @Override
+  public void validate(NoticeContainer noticeContainer) {
+    if (agencyTable.entityCount() <= 1) {
+      // fare_attributes.agency_id is not required when there is a single agency.
+      return;
```
Comment (briandonahue) [inline main/src/main/java/org/mobilitydata/gtfsvalidator/validator/FareAttributeAgencyIdValidator.java:49]: @isabelle-dr @aababilov I've added similar logic as part of #1318 and added a best practice [`MissingRecommendedFieldNotice`](https://github.com/MobilityData/gtfs-validator/pull/1318/files#diff-7d92ce3c768cd279570a721b435ef6247592eef8febe4f50687c05e52c6d2a8fR62) warning to recommend that `agency_id` be included even when there is only one agency.

**Comment by isabelle-dr** [comment]
Hello! Sorry it took a while. A few things here:
- `FeedHasLanguageValidator`
Is this mentioned in the spec somewhere? I couldn't find it. Adding this to the spec looks like a breaking change. Maybe something to add in the best Practice and flag as a warning here? Where is this coming from?
- `StationUsageValidator`
Reading the spec, I think stations without any children are okay? Please tell me if you think otherwise
- `BlockTripsWithConsistentTypeValidator`, `TransferFrequencyBasedTripValidator` and `TripShapeDistTraveledValidator`
These new validators make a lot of sense! That being said, so explicit spec mention. I suggest we first amend the spec and then merge them. Happy to propose the changes.
- `FareAttributeAgencyIdValidator`, `StopTimeInSequenceOrderValidator`, `StopLatLngRequiredValidator` -> LGTM
One question though: would `TripWithOutOfOrderShapeDistTraveledNotice` replace the existing `DecreassingOrEqualStopTimeDistance`?

**Comment by isabelle-dr** [comment]
I suggest we split this PR between what's explicitly mentioned in spec VS what is not.
What is explicitly mentioned: we just need the documentation update and we're good to merge.
What isn't: add the clear statements in the spec itself, and then merge

**Pull request closed without merging**